README.md
# gym-wordle
A wordle environment for openai/gym

# Installation
Install [openai-gym](https://gym.openai.com/docs/)

then install this package with
`pip install -e .`

# Usage
```
import gym 
import gym_wordle
env = gym.make('Wordle-v0')
```
See the [docs](https://gym.openai.com/docs/) for more info

# Environment details

This environment simulates a game of wordle using a wordle python clone from https://github.com/bellerb/wordle_solver

The action space is a discrete space of 12972 numbers which corresponds to a word from a list of all allowed wordle guesses and answers
The observation space is a dict with the guesses and colors for the current game.
Guesses is an array of `shape(5,6) #5 letters and 6 rows` where each element is a number from 0-26 where 0 is `''` and 26 is `z`
Colors is an array of the same shape, only each element is a number from 0-2 where 0 is a blank (or black or grey) square, 1 is a yellow square, and 2 is a green square

The reward calculation is as follows: 

The agent gets 1-6 points depending on how fast the agent guesses the word. For example, getting the word on the first guesses rewards 6 points, getting the word on the second guess rewards 5 points, etc

The agent also is rewarded for colors in the current row (so the current guess). 
Right now, the agent is rewarded 3 points for each green tile, and 1 point for each yellow tile. 
No points are giving for grey tiles


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.8.6 (2022-04-08)
### Fix
* Removed prints ([`89cf71e`](https://github.com/ryan-snyder/gym-wordle/commit/89cf71e15ca88dac1801c377ef3d0b7ab3458cf4))

## v0.8.5 (2022-04-08)
### Fix
* Fixed compute_reward ([`8b3b710`](https://github.com/ryan-snyder/gym-wordle/commit/8b3b710fd42df60fe85b7a9c96571f2bf114bf3c))

## v0.8.4 (2022-04-08)
### Fix
* Adding more print to try to fix error ([`9b29d67`](https://github.com/ryan-snyder/gym-wordle/commit/9b29d67f90649de614308cd04250a13a5e5c319e))

## v0.8.3 (2022-04-08)
### Fix
* Adding print for sanity check ([`f4bb349`](https://github.com/ryan-snyder/gym-wordle/commit/f4bb3498165690d958c5c81ee3853cae89e88b79))

## v0.8.2 (2022-04-08)
### Fix
* Adjusted compute_reward ([`2d38873`](https://github.com/ryan-snyder/gym-wordle/commit/2d388733bc7bbfc75f255a726033803adb333a01))

## v0.8.1 (2022-04-08)
### Fix
* Adjusting reward ([`4b9cd75`](https://github.com/ryan-snyder/gym-wordle/commit/4b9cd75081879242e8ac8e5f8c38fa530396dccc))

## v0.8.0 (2022-04-08)
### Feature
* Changing to new state ([`b1c0bdf`](https://github.com/ryan-snyder/gym-wordle/commit/b1c0bdf81eff5835ddb48412b03e5b71a362c8c1))

## v0.7.16 (2022-04-07)
### Fix
* Changing reward calc again ([`4b0de02`](https://github.com/ryan-snyder/gym-wordle/commit/4b0de02e54f23622f518e02c04f10004b23d90f3))

## v0.7.15 (2022-04-07)
### Fix
* Fixing error ([`ebcab27`](https://github.com/ryan-snyder/gym-wordle/commit/ebcab27c9d8f39e3d0825696d53dec3b8e88808b))

## v0.7.14 (2022-04-07)
### Fix
* Last reward adjustment ([`c00cd71`](https://github.com/ryan-snyder/gym-wordle/commit/c00cd717a4cff367f492cc96ca77b7e857a5eced))

## v0.7.13 (2022-04-07)
### Fix
* More reward adjustments ([`1c7fcd4`](https://github.com/ryan-snyder/gym-wordle/commit/1c7fcd42dcb3d4cecbd411f1e8f502156cf54dc1))

## v0.7.12 (2022-04-07)
### Fix
* Adjusting reward calc ([`4304ea9`](https://github.com/ryan-snyder/gym-wordle/commit/4304ea95ee6cb13e545a22e201271ffe9aa60bb9))

## v0.7.11 (2022-04-07)
### Fix
* Heavily punishing guessing same word ([`517074b`](https://github.com/ryan-snyder/gym-wordle/commit/517074ba69ba066b2131598563ac76d891e15274))

## v0.7.10 (2022-04-07)
### Fix
* Changing back to old compute_reward ([`89de522`](https://github.com/ryan-snyder/gym-wordle/commit/89de522c1de469bae3993ebb3da91ceb7327344b))

## v0.7.9 (2022-04-07)
### Fix
* Changing shape ([`6855940`](https://github.com/ryan-snyder/gym-wordle/commit/6855940b8d5282efbb9f25d9ad4d696c6a8dc004))

## v0.7.8 (2022-04-07)
### Fix
* Fixing her goals ([`331e467`](https://github.com/ryan-snyder/gym-wordle/commit/331e46776836f299748be922f310e7ec867bd0f3))

## v0.7.7 (2022-04-07)
### Fix
* Fixed compute_reward ([`b300801`](https://github.com/ryan-snyder/gym-wordle/commit/b30080139b028e0a1ccb9a4571b6178c41223686))

## v0.7.6 (2022-04-07)
### Fix
* Changed to random word but same space ([`0b01705`](https://github.com/ryan-snyder/gym-wordle/commit/0b01705be3d5d46d2d319c32265fadcdc01c8252))

## v0.7.5 (2022-04-07)
### Fix
* Removed extra print ([`17d26b9`](https://github.com/ryan-snyder/gym-wordle/commit/17d26b9288f9fdef11fb3c905d23376d71a761af))

## v0.7.4 (2022-04-07)
### Fix
* Adjusting rewards ([`b600437`](https://github.com/ryan-snyder/gym-wordle/commit/b6004376ab5095d6d6d6ed4db7d7b306f269e541))

## v0.7.3 (2022-04-07)
### Fix
* Fixed calculation error ([`e2f941c`](https://github.com/ryan-snyder/gym-wordle/commit/e2f941ca6323e6623849c4ce471fb998a7a6c504))

## v0.7.2 (2022-04-07)
### Fix
* Removing old reward stuff ([`1ee3f9f`](https://github.com/ryan-snyder/gym-wordle/commit/1ee3f9fbd90f49e4452c746ca90afc111bc34014))

## v0.7.1 (2022-04-07)
### Fix
* Removed extra prints ([`42d3e13`](https://github.com/ryan-snyder/gym-wordle/commit/42d3e1347cb6d092ec61ced34076e6639e576905))

## v0.7.0 (2022-04-07)
### Feature
* Adding more options to make stuff easier for agents ([`b51621d`](https://github.com/ryan-snyder/gym-wordle/commit/b51621dc7d1cb1b8d565ff72dedb796692051d75))

## v0.6.12 (2022-04-07)
### Fix
* Corrected action mask ([`36b707c`](https://github.com/ryan-snyder/gym-wordle/commit/36b707c019eef55c80f8e42a794d50e775520a1c))

## v0.6.11 (2022-04-07)
### Fix
* Fixed typo ([`28d0cce`](https://github.com/ryan-snyder/gym-wordle/commit/28d0cce97a9f51656c2395cf64708adcb8c12fe9))

## v0.6.10 (2022-04-07)
### Fix
* Fixed some errors and added action masking for maskable ppo ([`44847df`](https://github.com/ryan-snyder/gym-wordle/commit/44847dfb96c4247e411372d7140b3d1ccb1de3fb))

## v0.6.9 (2022-04-06)
### Fix
* Adju sted new reward for guessing the same word ([`6508212`](https://github.com/ryan-snyder/gym-wordle/commit/6508212d0b19df95b87b67e5a3d075a4412579a1))

## v0.6.8 (2022-04-06)
### Fix
* Add compute_reward ([`559cf46`](https://github.com/ryan-snyder/gym-wordle/commit/559cf461a0c5727a33c21aaee90cdf68491098c3))

## v0.6.7 (2022-04-06)
### Fix
* Fix spelling mistake ([`82d6900`](https://github.com/ryan-snyder/gym-wordle/commit/82d690002d1c1f0672234c3868862efa152f0a3e))

## v0.6.6 (2022-04-06)
### Fix
* Adjusted achieved goal and desired goal ([`e332979`](https://github.com/ryan-snyder/gym-wordle/commit/e33297914a34b0df627f6e83adf408d493eb475c))

## v0.6.5 (2022-04-06)
### Fix
* Removed print ([`598ce5a`](https://github.com/ryan-snyder/gym-wordle/commit/598ce5a1ba0e90c1fdd3eccf33ec763815d855b1))

## v0.6.4 (2022-04-06)
### Fix
* Fixed observation space ([`8786195`](https://github.com/ryan-snyder/gym-wordle/commit/8786195ac5e312a72fddc9a94dccd956f29194fa))

## v0.6.3 (2022-04-06)
### Fix
* Added back support for discrete action ([`55b15b0`](https://github.com/ryan-snyder/gym-wordle/commit/55b15b0d468880c98ad988f6b04d53720ee7289c))

## v0.6.2 (2022-04-06)
### Fix
* Fixed error ([`c816f4d`](https://github.com/ryan-snyder/gym-wordle/commit/c816f4d5c96d0b05fc0c5131077b1ced5eb6f6cd))

## v0.6.1 (2022-04-06)
### Fix

test.py
import gym
import os
import numpy as np
os.environ['SDL_VIDEODRIVER']='dummy'

import gym_wordle

env = gym.make('WordleEasy-v0', logging=True)
obs = env.reset()
print('Playing one game of wordle, with random action')
while not env.is_game_over:
    #env.render()
    obs, _, _, _ = env.step(env.action_space.sample())
    print(np.shape(obs['observation']))
    env.action_masks()
    print(env.compute_reward(obs['achieved_goal'], obs['desired_goal'], ''))
env.close()

gym_wordle/wordle.py
from copy import deepcopy

class Wordle:
    def __init__(self, word, rows=6, letters=5):
        self.g_count = 0
        self.word = word
        self.w_hash_table = {}
        if word is not None:
            for x, l in enumerate(word):
                if l in self.w_hash_table:
                    self.w_hash_table[l]['count'] += 1
                    self.w_hash_table[l]['pos'].append(x)
                else:
                    self.w_hash_table[l] = {'count':1, 'pos':[x]}
        self.rows = rows
        self.letters = letters
        self.board = [['' for _ in range(letters)] for _ in range(rows)]
        self.colours = [['B' for _ in range(letters)] for _ in range(rows)]
        self.alph = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']

    def is_end(self):
        if self.board[-1] != ['' for _ in range(self.letters)]:
            return True
        else:
            r = self.game_result()
            if r[0] == True:
                return True
            else:
                return False

    def game_result(self):
        win = (False, 100)
        for i, r in enumerate(self.board):
            if self.word == ''.join(r):
                win = (True, i)
                break
        return win

    def update_board(self, u_inp):
        w_hash_table = deepcopy(self.w_hash_table)
        i_hash_table = {}
        for x, l in enumerate(str(u_inp).upper()):
            self.board[self.g_count][x] = l
            if l in i_hash_table:
                i_hash_table[l].append(x)
            else:
                i_hash_table[l] = [x]
        colours = {'G':[],'B':[],'Y':[]}
        for l in i_hash_table:
            if l in w_hash_table:
                g_hold = []
                for p in i_hash_table[l]:
                    if p in w_hash_table[l]['pos']:
                        g_hold.append(p)
                for p in g_hold:
                    i_hash_table[l].remove(p)
                colours['G'] += g_hold
                if len(g_hold) < w_hash_table[l]['count']:
                    y_hold = []
                    for p in i_hash_table[l]:
                        y_hold.append(p)
                        if len(y_hold) == w_hash_table[l]['count']:
                            break
                    for p in y_hold:
                        i_hash_table[l].remove(p)
                    colours['Y'] += y_hold
                for p in i_hash_table[l]:
                    colours['B'].append(p)
            else:
                colours['B'] += i_hash_table[l]
                i_hash_table[l] = []
        for c in colours:
            for p in colours[c]:
                self.colours[self.g_count][p] = c
        self.g_count += 1

    def valid_guess(self, u_inp):
        if len(u_inp) == 5 and False not in [False for s in str(u_inp).upper() if s not in self.alph]:
            return True
        else:
            return False

gym_wordle/envs/state.py
"""
Keep the state in a 1D int array
index[0] = remaining steps
Rest of data is laid out as binary array
[1..27] = whether char has been guessed or not
[[status, status, status, status, status]
 for _ in "ABCD..."]
where status has codes
 [1, 0, 0] - char is definitely not in this spot
 [0, 1, 0] - char is maybe in this spot
 [0, 0, 1] - char is definitely in this spot
"""
import collections
from typing import List
import numpy as np


WORDLE_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
WORDLE_N = 5

WordleState = np.ndarray


def get_nvec(max_turns: int):
    return [max_turns] + [2] * len(WORDLE_CHARS) + [2] * 3 * WORDLE_N * len(WORDLE_CHARS)


def new(max_turns: int) -> WordleState:
    return np.array(
        [max_turns] + [0] * len(WORDLE_CHARS) + [0, 1, 0] * WORDLE_N * len(WORDLE_CHARS),
        dtype=np.int32)


def remaining_steps(state: WordleState) -> int:
    return state[0]


NO = 0
SOMEWHERE = 1
YES = 2


def update_from_mask(state: WordleState, word: str, mask: List[int]) -> WordleState:
    """
    return a copy of state that has been updated to new state
    From a mask we need slighty different logic since we don't know the
    goal word.
    :param state:
    :param word:
    :param goal_word:
    :return:
    """
    state = state.copy()

    prior_yes = []
    prior_maybe = []
    # We need two passes because first pass sets definitely yesses
    # second pass sets the no's for those who aren't already yes
    state[0] -= 1
    for i, c in enumerate(word):
        cint = ord(c) - ord(WORDLE_CHARS[0])
        offset = 1 + len(WORDLE_CHARS) + cint * WORDLE_N * 3
        state[1 + cint] = 1
        if mask[i] == YES:
            prior_yes.append(c)
            # char at position i = yes, all other chars at position i == no
            state[offset + 3 * i:offset + 3 * i + 3] = [0, 0, 1]
            for ocint in range(len(WORDLE_CHARS)):
                if ocint != cint:
                    oc_offset = 1 + len(WORDLE_CHARS) + ocint * WORDLE_N * 3
                    state[oc_offset + 3 * i:oc_offset + 3 * i + 3] = [1, 0, 0]

    for i, c in enumerate(word):
        cint = ord(c) - ord(WORDLE_CHARS[0])
        offset = 1 + len(WORDLE_CHARS) + cint * WORDLE_N * 3
        if mask[i] == SOMEWHERE:
            prior_maybe.append(c)
            # Char at position i = no, other chars stay as they are
            state[offset + 3 * i:offset + 3 * i + 3] = [1, 0, 0]
        elif mask[i] == NO:
            # Need to check this first in case there's prior maybe + yes
            if c in prior_maybe:
                # Then the maybe could be anywhere except here
                state[offset+3*i:offset+3*i+3] = [1, 0, 0]
            elif c in prior_yes:
                # No maybe, definitely a yes, so it's zero everywhere except the yesses
                for j in range(WORDLE_N):
                    # Only flip no if previously was maybe
                    if state[offset + 3 * j:offset + 3 * j + 3][1] == 1:
                        state[offset + 3 * j:offset + 3 * j + 3] = [1, 0, 0]
            else:
                # Just straight up no
                state[offset:offset+3*WORDLE_N] = [1, 0, 0]*WORDLE_N

    return state


def get_mask(word: str, goal_word: str) -> List[int]:
    # Definite yesses first
    mask = [0, 0, 0, 0, 0]
    counts = collections.Counter(goal_word)
    for i, c in enumerate(word):
        if goal_word[i] == c:
            mask[i] = 2
            counts[c] -= 1

    for i, c in enumerate(word):
        if mask[i] == 2:
            continue
        elif c in counts:
            if counts[c] > 0:
                mask[i] = 1
                counts[c] -= 1
            else:
                for j in range(i+1, len(mask)):
                    if mask[j] == 2:
                        continue
                    mask[j] = 0

    return mask

def update_mask(state: WordleState, word: str, goal_word: str) -> WordleState:
    """
    return a copy of state that has been updated to new state
    :param state:
    :param word:
    :param goal_word:
    :return:
    """
    mask = get_mask(word, goal_word)
    return update_from_mask(state, word, mask)


def update(state: WordleState, word: str, goal_word: str) -> WordleState:
    state = state.copy()
    state[0] -= 1
    for i, c in enumerate(word.upper()):
        cint = ord(c) - ord(WORDLE_CHARS[0])
        offset = 1 + len(WORDLE_CHARS) + cint * WORDLE_N * 3
        state[1 + cint] = 1
        if goal_word.upper()[i] == c:
            # char at position i = yes, all other chars at position i == no
            state[offset + 3 * i:offset + 3 * i + 3] = [0, 0, 1]
            for ocint in range(len(WORDLE_CHARS)):
                if ocint != cint:
                    oc_offset = 1 + len(WORDLE_CHARS) + ocint * WORDLE_N * 3
                    state[oc_offset + 3 * i:oc_offset + 3 * i + 3] = [1, 0, 0]
        elif c in goal_word.upper():
            # Char at position i = no, other chars stay as they are
            state[offset + 3 * i:offset + 3 * i + 3] = [1, 0, 0]
        else:
            # Char at all positions = no
            state[offset:offset + 3 * WORDLE_N] = [1, 0, 0] * WORDLE_N

    return state

gym_wordle/envs/wordle_env.py
import gym
from gym import error, spaces, utils
from gym.utils import seeding
import pandas as pd
from gym_wordle.wordle import Wordle
import pygame
import numpy as np
import os

class WordleEnv(gym.Env):
    metadata = {'render.modes': ['human']}
    SCREEN_DIM = 500
    GREEN = "#6aaa64"
    YELLOW = "#c9b458"
    GREY = "#787c7e"
    OUTLINE = "#d3d6da"
    FILLED_OUTLINE = "#878a8c"

    def __init__(self, logging=False):
        current_dir = os.path.dirname(os.path.realpath(__file__))
        self.logging = logging
        self.answers = pd.read_csv('{}/wordle-answers-alphabetical.txt'.format(current_dir), header=None, names=['words'])
        self.screen = None
        self.isopen = False
        self.GUESSES = 6
        self.LETTERS = 5
        self.WORD = self.answers['words'].sample(n=1).tolist()[0].upper()
        self.WORDLE = Wordle(self.WORD, self.GUESSES, self.LETTERS)
        self.alpha = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']
        self.colors = ['B', 'Y', 'G']
        self.is_game_over = False
        self.guessed_words = []
        self.blank_letters = []

        file_names = ['{}/wordle-allowed-guesses.txt'.format(current_dir), '{}/wordle-answers-alphabetical.txt'.format(current_dir)]
        self.word_bank = pd.concat((pd.read_csv(f, header=None, names=['words']) for f in file_names), ignore_index=True).sort_values('words')['words'].tolist()
        # our action space is the total amount of possible words to guess
        self.action_space = spaces.Discrete(12972)
        #our observation space is the current wordle board in form of (letter, color) with 5x6 (5 letters, 6 guesses)
        #modified to work with gym/baselines
        #same thing basically, only 0-26 is '' to z and 27-29 is B, Y, G
        # first 6 rows are guesses and last 6 rows are colors
        # changed shape to be 3 dimensions so that we can apply conv2d layers to it
        # at some point we should try to normalize the obs space
        # since right now its on a 0-29 scale instead of a 0-1.
        self.observation_space = spaces.Box(low=0, high=29, shape=(1,12,5), dtype='int32')
        self.current_episode = -1
        self.episode_memory: List[Any] = []

    def step(self, action):
        if self.is_game_over:
            return RuntimeError('Episode is already done')
        self._take_action(action)
        reward = self._get_reward()
        observation = self._get_observation()
        return observation, reward, self.is_game_over, {}

    def reset(self):
        self.current_episode = -1
        self.episode_memory.append([])
        self.is_game_over = False
        self.WORD = self.answers['words'].sample(n=1).tolist()[0].upper()
        self.WORDLE = Wordle(self.WORD, self.GUESSES, self.LETTERS)
        self.guessed_words = []
        self.blank_letters = []
        if self.logging:
            print(self.WORDLE.word)
        self.close()
        return self._get_observation()

    def render(self, mode='human'):
        if self.screen is None:
            pygame.init()
            pygame.display.init()
            self.screen = pygame.display.set_mode((self.SCREEN_DIM, self.SCREEN_DIM))
        font = pygame.font.Font('freesansbold.ttf', 30)
        for col in range(0, 5):
            for row in range(0, 6):
                pygame.draw.rect(self.screen, self.OUTLINE, [col * 100 + 12, row * 100 + 12, 75, 75], 3, 5)
                color = self.GREEN if self.WORDLE.colours[row][col] == 'G' else self.YELLOW if self.WORDLE.colours[row][col] == 'Y' else self.GREY
                piece_text = font.render(self.WORDLE.board[row][col], True, color)
                self.screen.blit(piece_text, (col * 100 + 30, row * 100 + 25))
        #pygame.draw.rect(screen, self.GREEN, [5, turn * 100 + 5, WIDTH - 10, 90], 3, 5)
        if mode == "human":
            pygame.event.pump()
            pygame.display.flip()             
    def close(self):
        if self.screen is not None:
            pygame.display.quit()
            pygame.quit()
            self.isopen = False

    def _take_action(self, action):
        # turn action into guess
        guess = self.word_bank[action]
        self.episode_memory[self.current_episode].append(guess)
        self.guessed_words.append(guess)
        if self.logging:
            print(guess)
        self.WORDLE.update_board(guess)
        res = self.WORDLE.colours[self.WORDLE.g_count-1]
        self.blank_letters.extend([ l for i,l in enumerate(guess) if res[i] == 'B' and l not in self.blank_letters])
        self.is_game_over = self.WORDLE.word == guess or self.WORDLE.g_count == self.GUESSES

    def _get_reward(self):
        result, tries = self.WORDLE.game_result()
        rewards = np.zeros(5)
        #heavily penealize guessing the same word multiple times
        #If a word isn't the right guess, we shouldn't guess it again
        #could do the same thing for letters, as if a letter is blank(grey)
        # then the only reason to use a word with a letter in it
        # is to check other letter posistions
        #so it shouldn't be a heavy penalty but it should be a penalty
        for i,c in enumerate(self.WORDLE.colours[self.WORDLE.g_count-1]):
            if c == self.colors[2]:
                rewards[i] = 2
            elif c == self.colors[1]:
                rewards[i] = 1
        #check guesses up to and including our current guess
        if self.logging:
            print(self.WORD)
            print(rewards)
        reward = np.mean(rewards)
        for g in range(self.WORDLE.g_count):
            word = self.WORDLE.board[g]
            current = ''.join(word)
            if current in self.guessed_words:
                return 0
            for l in word: 
                if l in self.blank_letters:
                    reward -= 0.3
        return reward

    def _get_observation(self):
        board = np.array(self.WORDLE.board) #2d array of 5x6
        colors = np.array(self.WORDLE.colours) #2d array of 5x6
        results = np.vstack((board, colors)) #stacks boards and colors by rows resulting in a 2d array of 5x12
        convertletterstonum = lambda letter: [self.alpha.index(l) + 1 if l in self.alpha else 0 for l in letter]
        convertcolortonum = lambda color: [self.colors.index(c)+27 for c in color]
        guesses = np.array([convertletterstonum(l) if i <=5 else convertcolortonum(l) for i, l in enumerate(results)])
        guesses3d = np.expand_dims(guesses, axis=0)
        if self.logging:
            print(np.shape(guesses))
            print(np.shape(guesses3d))
        return guesses3d


gym_wordle/envs/wordle_env_easy.py
import gym
from gym import error, spaces, utils
from gym.utils import seeding
import pandas as pd
from gym_wordle.wordle import Wordle
import pygame
import numpy as np
import os
import gym_wordle.envs.state as state

class WordleEnvEasy(gym.Env):
    metadata = {'render.modes': ['human']}
    SCREEN_DIM = 500
    GREEN = "#6aaa64"
    YELLOW = "#c9b458"
    GREY = "#787c7e"
    OUTLINE = "#d3d6da"
    FILLED_OUTLINE = "#878a8c"

    def __init__(self, logging=False, action_type='Discrete', words=100, random=False):
        current_dir = os.path.dirname(os.path.realpath(__file__))
        self.logging = logging
        self.random = random
        self.screen = None
        self.isopen = False
        self.GUESSES = 6
        self.LETTERS = 5
        if random:
            print('using random solution set and random solution, of size: ', words)
            self.answers = pd.read_csv('{}/wordle-answers-alphabetical.txt'.format(current_dir), header=None, names=['words']).sample(n=words, ignore_index=True)
            self.WORD = self.answers['words'].sample(n=1).tolist()[0].upper()
        else:
            print('using same solution set and random solution, of size: ', words)
            self.answers = pd.read_csv('{}/wordle-answers-alphabetical.txt'.format(current_dir), header=None, names=['words']).sort_values('words').head(words)
            self.WORD = self.answers['words'].sample(n=1).tolist()[0].upper()
        self.WORDLE = Wordle(self.WORD, self.GUESSES, self.LETTERS)
        self.alpha = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']
        self.colors = ['B', 'Y', 'G']
        self.is_game_over = False
        self.guessed_words = []
        self.blank_letters = []
        self.rewards = []
        file_names = ['{}/wordle-answers-alphabetical.txt'.format(current_dir)]
        #self.word_bank = pd.concat((pd.read_csv(f, header=None, names=['words']) for f in file_names), ignore_index=True).sort_values('words')
        self.word_bank = self.answers
        # our action space is the total amount of possible words to guess
        if action_type == 'Discrete':
            self.action_space = spaces.Discrete(words)
        else:
        # TODO: change this to provide x guesses, and choose the one which scores the highest?
        # or is that too forgiving
            self.action_space = spaces.Box(low=0, high=2315, shape=(1,), dtype='int32')
        #our observation space is the current wordle board in form of (letter, color) with 5x6 (5 letters, 6 guesses)
        #modified to work with gym/baselines
        #same thing basically, only 0-26 is '' to z and 27-29 is B, Y, G
        # first 6 rows are guesses and last 6 rows are colors
        # changed shape to be 3 dimensions so that we can apply conv2d layers to it
        # at some point we should try to normalize the obs space
        # since right now its on a 0-29 scale instead of a 0-1.
        #self.observation_space = spaces.Box(low=0, high=29, shape=(1,12,5), dtype='int32')
        self.observation_space = spaces.Dict({
            'observation': spaces.Box(low=0, high=6, shape=(417,), dtype='int32'),
            'achieved_goal': spaces.Box(low=27, high=29, shape=(5,), dtype='int32'),
            'desired_goal': spaces.Box(low=29, high=29, shape=(5,), dtype='int32')
        })
        self.current_episode = -1
        self.episode_memory: List[Any] = []
        self.state: state.WordleState = None
        self.state_updater = state.update

    def step(self, action):
        if self.is_game_over:
            return RuntimeError('Episode is already done')
        self._take_action(action)
        reward = 0
        observation = self._get_observation()
        if self.word_bank['words'].to_list()[self.current_guess] == self.WORD.lower():
            self.is_game_over = True
        elif self.WORDLE.g_count == self.GUESSES:
            self.is_game_over = True
        reward += self._get_reward()
        self.rewards.append(reward)
        if self.logging:
            print(self.WORD)
            print(self.guessed_words)
            print(self.rewards)
        return observation, reward, self.is_game_over, {}

    def reset(self):
        self.current_episode = -1
        self.episode_memory.append([])
        self.is_game_over = False
        if self.random:
            self.WORD = self.answers.loc[:,'words'].sample(n=1).tolist()[0].upper()
        else:
            self.WORD = self.answers['words'].sample(n=1).tolist()[0].upper()
        self.WORDLE = Wordle(self.WORD, self.GUESSES, self.LETTERS)
        self.guessed_words = []
        self.blank_letters = []
        self.rewards = []
        if self.logging:
            print(self.WORDLE.word)
        self.close()
        self.state = state.new(self.GUESSES)
        return self._get_observation()

    def render(self, mode='human'):
        if self.screen is None:
            pygame.init()
            pygame.display.init()
            self.screen = pygame.display.set_mode((self.SCREEN_DIM, self.SCREEN_DIM))
        font = pygame.font.Font('freesansbold.ttf', 30)
        for col in range(0, 5):
            for row in range(0, 6):
                pygame.draw.rect(self.screen, self.OUTLINE, [col * 100 + 12, row * 100 + 12, 75, 75], 3, 5)
                color = self.GREEN if self.WORDLE.colours[row][col] == 'G' else self.YELLOW if self.WORDLE.colours[row][col] == 'Y' else self.GREY
                piece_text = font.render(self.WORDLE.board[row][col], True, color)
                self.screen.blit(piece_text, (col * 100 + 30, row * 100 + 25))
        #pygame.draw.rect(screen, self.GREEN, [5, turn * 100 + 5, WIDTH - 10, 90], 3, 5)
        if mode == "human":
            pygame.event.pump()
            pygame.display.flip()             
    def close(self):
        if self.screen is not None:
            pygame.display.quit()
            pygame.quit()
            self.isopen = False

    def _take_action(self, action):
        # turn action into guess
        guess = self.word_bank['words'].to_list()[action]
        self.episode_memory[self.current_episode].append(guess)
        self.guessed_words.append(guess)
        self.current_guess = action
        self.WORDLE.update_board(guess)
        res = self.WORDLE.colours[self.WORDLE.g_count-1]
        for i,l in enumerate(guess):
            if res[i] == 'B' and l not in self.blank_letters:
                self.blank_letters.append(l)
        if self.WORDLE.word.lower() == guess:
            print('~~~~~~AGENT GOT IT RIGHT~~~~~~')
            print(self.guessed_words)
    def _get_reward(self):
        new_reward = 0
        if self.word_bank['words'].to_list()[self.current_guess] == self.WORD.lower():
            if self.WORDLE.g_count > 1:
                new_reward = 10
            else: new_reward = -20
        elif self.WORDLE.g_count == self.GUESSES:
            new_reward = -10
        return new_reward
    def action_masks(self):
        action_mask = [w in self.guessed_words for w in self.word_bank['words'].tolist()]
        #action_mask = [self.guessed_words[key]['action'] for key in self.guessed_words.keys()]
        return action_mask
    # TODO: adjust get reward and compute reward to take into account the desired goal
    # But i think this is fine for right now, since our _get_reward does take into account our desired goal
    def compute_reward(self, achieved_goal, desired_goal, info):
        rewards = np.zeros((len(achieved_goal)))
        for i,result in enumerate(achieved_goal):
            for goal in desired_goal:
                adjusted = 10
                for g in range(len(goal)):
                    adjusted -= goal[g] - result[g]
                    rewards[i] = adjusted
        return np.mean(rewards)

    def _get_observation(self):
        board = np.array(self.WORDLE.board) #2d array of 5x6
        colors = np.array(self.WORDLE.colours) #2d array of 5x6
        results = np.vstack((board, colors)) #stacks boards and colors by rows resulting in a 2d array of 5x12
        convertletterstonum = lambda letter: [self.alpha.index(l) + 1 if l in self.alpha else 0 for l in letter]
        convertcolortonum = lambda color: [self.colors.index(c)+27 for c in color]
        guesses = np.array([convertletterstonum(l) if i <=5 else convertcolortonum(l) for i, l in enumerate(results)])
        colors = np.array(convertcolortonum(colors[self.WORDLE.g_count-1]))
        #guesses3d = np.expand_dims(guesses, axis=0)
        if self.state_updater != None and self.WORDLE.g_count != 0:
            return { 
                'observation': self.state_updater(state=self.state, word=self.guessed_words[self.WORDLE.g_count-1], goal_word=self.WORD.lower()), 
                'achieved_goal': colors, 'desired_goal': [29,29,29,29,29]
            }
        elif self.state_updater != None: 
            return {
                'observation': state.get_nvec(6),
                'achieved_goal': colors, 'desired_goal': [29,29,29,29,29]
            }
        return { 'observation': guesses, 'achieved_goal': colors, 'desired_goal': [29,29,29,29,29]}
